Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,625 @@
+            return self.channelIsNotWritable()
+        }
+    }
+
+    private mutating func channelIsWritable() -> Action {
+        self.isChannelWritable = true
+
+        switch self.state {
+        case .initialized,
+             .verifyingRequest,
+             .running(.streaming(_, _, producer: .producing), _),
+             .running(.endSent, _),
+             .finished,
+             .failed:
+            return .wait
+
+        case .waitForChannelToBecomeWritable(let head):
+            return self.startSendingRequestHead(head)
+
+        case .running(.streaming(_, _, producer: .paused), .receivingBody(let head, _)) where head.status.code >= 300:
+            // If we are receiving a response with a status of >= 300, we should not send out
+            // further request body parts. The remote already signaled with status >= 300 that it
+            // won't be interested. Let's save some bandwidth.
+            return .wait
+
+        case .running(.streaming(let expectedBody, let sentBodyBytes, producer: .paused), let responseState):
+            let requestState: RequestState = .streaming(
+                expectedBodyLength: expectedBody,
+                sentBodyBytes: sentBodyBytes,
+                producer: .producing
+            )
+
+            self.state = .running(requestState, responseState)
+            return .resumeRequestBodyStream
+        }
+    }
+
+    private mutating func channelIsNotWritable() -> Action {
+        self.isChannelWritable = false
+
+        switch self.state {
+        case .initialized,
+             .verifyingRequest,
+             .waitForChannelToBecomeWritable,
+             .running(.streaming(_, _, producer: .paused), _),
+             .running(.endSent, _),
+             .finished,
+             .failed:
+            return .wait
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, producer: .producing), let responseState):
+            let requestState: RequestState = .streaming(
+                expectedBodyLength: expectedBodyLength,
+                sentBodyBytes: sentBodyBytes,
+                producer: .paused
+            )
+            self.state = .running(requestState, responseState)
+            return .pauseRequestBodyStream
+        }
+    }
+
+    mutating func readEventCaught() -> Action {
+        switch self.state {
+        case .initialized,
+             .verifyingRequest,
+             .waitForChannelToBecomeWritable,
+             .running(_, .waitingForHead),
+             .running(_, .endReceived),
+             .finished,
+             .failed:
+            // If we are not in the middle of streaming the response body, we always want to get
+            // more data...
+            return .read
+        case .running(_, .receivingBody(_, .downstreamIsConsuming(readPending: true))):
+            preconditionFailure("It should not be possible to receive two reads after each other, if the first one hasn't been forwarded.")
```

Comment:
Why is that?

Reply:
Given that reads are triggered after a `readComplete` (we don't modify any channel options), I would assume that a new read event can only happen, if in the mean time data was actually read from the wire. To read from the wire the consumer needs to signal demand: this will transform `.downstreamIsConsuming(readPending: true)` -> `.downstreamHasDemand`. Once we get new data: `.downstreamIsConsuming(readPending: false)`, only then we should get into a state where a new `read` event might happen.

Reply:
A `read` event is not required to trigger any non-zero number of calls to `channelRead`: it's only required to trigger a single call to `channelReadComplete`. More broadly, there's no real reason to forbid this, I don't think.